hail/src/main/scala/is/hail/expr/ir/NormalizeNames.scala

```diff
@@ -0,0 +1,79 @@
+
+    ir match {
+      case Let(name, value, body) =>
+        val newName = gen()
+        Let(newName, normalize(value), normalize(body, env.bind(name, newName)))
+      case Ref(name, typ) =>
+        Ref(env.lookup(name), typ)
+      case AggLet(name, value, body) =>
+        val newName = gen()
+        AggLet(newName, normalize(value), normalize(body, env, Some(aggEnv.get.bind(name, newName))))
+      case ArraySort(a, left, right, compare) =>
+        val newLeft = gen()
+        val newRight = gen()
+        ArraySort(normalize(a), newLeft, newRight, normalize(compare, env.bind(left -> newLeft, right -> newRight)))
+      case ArrayMap(a, name, body) =>
+        val newName = gen()
+        ArrayMap(normalize(a), newName, normalize(body, env.bind(name, newName)))
+      case ArrayFilter(a, name, body) =>
+        val newName = gen()
+        ArrayFilter(normalize(a), newName, normalize(body, env.bind(name, newName)))
+      case ArrayFlatMap(a, name, body) =>
+        val newName = gen()
+        ArrayFlatMap(normalize(a), newName, normalize(body, env.bind(name, newName)))
+      case ArrayFold(a, zero, accumName, valueName, body) =>
+        val newAccumName = gen()
+        val newValueName = gen()
+        ArrayFold(normalize(a), normalize(zero), newAccumName, newValueName, normalize(body, env.bind(accumName -> newAccumName, valueName -> newValueName)))
+      case ArrayScan(a, zero, accumName, valueName, body) =>
+        val newAccumName = gen()
+        val newValueName = gen()
+        ArrayScan(normalize(a), normalize(zero), newAccumName, newValueName, normalize(body, env.bind(accumName -> newAccumName, valueName -> newValueName)))
+      case ArrayFor(a, valueName, body) =>
+        val newValueName = gen()
+        ArrayFor(normalize(a), newValueName, normalize(body, env.bind(valueName, newValueName)))
+      case ArrayAgg(a, name, body) =>
+        assert(aggEnv.isEmpty)
+        val newName = gen()
+        ArrayAgg(normalize(a), newName, normalize(body, env, Some(env.bind(name, newName))))
+      case ArrayLeftJoinDistinct(left, right, l, r, keyF, joinF) =>
+        val newL = gen()
+        val newR = gen()
+        val newEnv = env.bind(l -> newL, r -> newR)
+        ArrayLeftJoinDistinct(normalize(left), normalize(right), newL, newR, normalize(keyF, newEnv), normalize(joinF, newEnv))
+      case AggExplode(a, name, aggBody) =>
+        val newName = gen()
+        AggExplode(normalize(a, aggEnv.get, None), newName, normalize(aggBody, env, Some(aggEnv.get.bind(name, newName))))
+      case AggArrayPerElement(a, name, aggBody) =>
+        val newName = gen()
+        AggArrayPerElement(normalize(a, aggEnv.get, None), newName, normalize(aggBody, env, Some(aggEnv.get.bind(name, newName))))
+      case ApplyAggOp(ctorArgs, initOpArgs, seqOpArgs, aggSig) =>
+        ApplyAggOp(ctorArgs.map(a => normalize(a)),
+          initOpArgs.map(_.map(a => normalize(a))),
```

Comment:
How do you know if you got all of the cases? Did you use the parser code? I looked through each, but I don't think I'll catch omission errors.

Reply:
I searched the IR.scala and looked for every occurrence of `String`.  I'm working on @tpoterba on abstractions for getting binding information about IR.  It's not quite clear how to do that right, but this use case should definitely be considered.

If I miss one, the common case is that I won't put it in the env, and a Let that references it will throw an assert error when it calls env.lookup.